doc/netcomics.1
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
''' Revision 1.1  2000-06-12 11:47:31  hochstrb
''' Initial revision
'''
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH NETCOMICS 1 "netcomics-0.12" "12/Jun/100" "Web Utilities"
.IX Title "NETCOMICS 1"
.UC
.IX Name "netcomics - retrieve comics from the Internet"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
netcomics \- retrieve comics from the Internet
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\fBnetcomics\fR [\fB\-bBDhiIlLosuvv\fR] [\fB\-c,\-C\fR \fI"comic ids\*(R"\fR] [\fB\-p\fR \fIproxy\fR] [\fB\-S,\-T,\-E\fR \fIdate\fR]
                 [\fB\-n,\-N\fR \fIdays\fR] [\fB\-d,\-m,\-t\fR \fIdir\fR] [\fB\-f\fR \fIdate_fmt\fR] [\fB\-g\fR [\fIprogram\fR]] [\fB\-nD\fR]
                 [\fB\-r\fR \fIrc_file\fR] [\fB\-W,\-w\fR[=\fIn\fR]] [\fB\-nw\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fInetcomics\fR program will download today's comic strips from
the Web, and place them in a spool directory where they can be
retrieved for display.  Because each website that carries comic
strips chooses how old of strips to show, the comic strips downloaded
will actually be from different dates.  Most common will be comic
strips that are 1 week, 2 weeks, and 3 weeks old. Also, each website
is not updated at the same time, nor are any of them updated at a
consistent time during the day.  Therefore, when running netcomics
as a cron job in the early morning, you may need to rerun it by hand
a little later in the day to get the comics that it couldn't find.
The exact command to run is given at the end of \fInetcomics\fR output
if any failures occurred.
.PP
\fInetcomics\fR also supports retrieving specific dates of comics.  A
\*(L"Starting Date\*(R" may be specified with \fB\-S\fR.  An \*(L"Ending Date\*(R" may be
specified with \fB\-E\fR.  And a specific date may be specified with
\fB\-T\fR.  All dates are given in the M\-D-[YY]YY format.  If a start date
is given without an end date, \fB\-n\fR may be used to specify the number
of days of comics to retrieve starting at the specified start date.
If an end date is given without a start date, \fB\-n\fR will specify the
number of days of comics to retrieve counting backwards from the end
date.  If a start date is given without \fB\-n\fR or \fB\-E\fR, then the end
date is assumed to be today's date.  If the option given to \fB\-E\fR,
\fB\-S\fR, or \fB\-T\fR is just an integer, that integer is interpreted as
the number of days prior to the current day (specify specific dates
relatively).
.PP
Another way to specify the dates to download comics, is to use \fB\-N\fR
in conjunction with \fB\-n\fR.  The argument to \fB\-N\fR is the number of
days prior to the current day to retrieve comics.  Note that this does
not specify an actual date.  It rather indicates the number of days
ago a comic was made available, rather than the actual date of the
comic.  To see how far behind the date of each comic is from the date
it actually gets released, use \fB\-l\fR.  So if today is Wednesday, and you
specify \fB\-N 2\fR, \fInetcomics\fR will download the comics that were made
available on Monday.  This is useful for running netcomics in a
timezone that is way ahead of the timezones the comics\*(R' websites are
located.
.PP
\fInetcomics\fR was created for the purpose of giving your weary mind a 
little relief from your hectic workday, so another script
called \fIshow_comics\fR has been provided to periodically popup the 
retrieved comic strips throughout the workday.  Another script called
\fIdisplay_comics\fR has also been provided, but it is only an example
script of what can be done for displaying comics on many users
computers on a network.
.PP
\fBImportant:\fR  The further east your timezone is from the US, the later 
in the day you'll have to run \fInetcomics\fR.  As a reference, I suggest
those whose timezone in GMT to wait to run the script at 12:30pm if
they want to get all the comics at a time that's pretty likely to have
had all of the websites updated.  Use \fB\-N\fR if you want to run the
script early in the morning and are having problems getting comics to
download. Also, just because a comic failed to download doesn't mean
that the module for that comic is broken\*(--it most likely the website
just hasn't been updated yet.
.PP
\fInetcomics\fR can also create an HTML file, \*(L"index.html\*(R", in the
directory you have the comics placed.   If a number is given with
\fB\-W\fR or \fB\-w\fR, it will be used to determine the number of comics to
be placed in each html file.  Subsequent files are named \*(L"comic#.html\*(R".
.PP
\fBDisclaimer:\fR Do not put the comics up on the Internet!  You should
only use them for your own use.  Also, do not redistribute the comics
downloaded by \fInetcomics\fR in any other way unless you receive written
authorization from each publisher.
.SH "OPTIONS"
.IX Header "OPTIONS"
.Ip "\fB\-b\fR" 4
.IX Item "\fB\-b\fR"
Get black & white comics if their is a choice between that and color.
.Ip "\fB\-B\fR" 4
.IX Item "\fB\-B\fR"
Don't get black & white comics if their is a choice between that and color.
Use this option to override the setting in the rc file.
.Ip "\fB\-c\fR \fI'comic_ids\*(R'\fR" 4
.IX Item "\fB\-c\fR \fI'comic_ids\*(R'\fR"
Get the supplied comics (ids are separated by white spaces).
This option may be repeated, and may not be used in conjunction with \fB\-C\fR.
.Ip "\fB\-C\fR \fI'comic_ids\*(R'\fR" 4
.IX Item "\fB\-C\fR \fI'comic_ids\*(R'\fR"
Don't get the supplied comics (ids are separated by white spaces).
This option may be repeated, and may not be used in conjunction with \fB\-c\fR.
.Ip "\fB\-d\fR \fIdir\fR" 4
.IX Item "\fB\-d\fR \fIdir\fR"
Put comics into directory. Default is /home/benh/www/comics.
.Ip "\fB\-D\fR" 4
.IX Item "\fB\-D\fR"
Delete files in directory before retrieving.
.Ip "\fB\-E\fR [\fIdate\fR | \fIdays\fR]" 4
.IX Item "\fB\-E\fR [\fIdate\fR | \fIdays\fR]"
Specify an ending date, or the date that is the specified number of
days prior to today, with which to define the range of days of comics
to retrieve.  Must be used in combination with one of \fB\-S\fR or \fB\-n\fR.
The date is of the form: M\-D-Y.
.Ip "\fB\-f\fR \fIdate_fmt\fR" 4
.IX Item "\fB\-f\fR \fIdate_fmt\fR"
Specify the date format used when naming files.  Default is \f(CW'%y%m%d'\fR.
.Ip "\fB\-g\fR [\fIprogram\fR]" 4
.IX Item "\fB\-g\fR [\fIprogram\fR]"
Specify a program to use to perform \s-1HTTP\s0 requests.  The \s-1URL\s0 to retrieve 
is attached to the end of the provided text without a space. The 
program must write the retrieved content of the \s-1URL\s0 on its stdout.
The default command is:
.Sp
.Vb 1
\&   'wget -q -O - '.
.Ve
.Ip "\fB\-h\fR" 4
.IX Item "\fB\-h\fR"
Show usage. Comics will not be downloaded.
.Ip "\fB\-i\fR Don't create an index for the webpages." 4
.IX Item "\fB\-i\fR Don't create an index for the webpages."
.Ip "\fB\-I\fR Create an index for the webpages. Specify this to override the setting  in the rc file if need be." 4
.IX Item "\fB\-I\fR Create an index for the webpages. Specify this to override the setting  in the rc file if need be."
.Ip "\fB\-l\fR" 4
.IX Item "\fB\-l\fR"
List supported comics & their identifiers, sorted by name alphabetically.
Comics will not be downloaded.
.Ip "\fB\-L\fR" 4
.IX Item "\fB\-L\fR"
Sort comics in webpages by date and when using \fB\-l\fR, by days behind.
.Ip "\fB\-m\fR \fIdir\fR" 4
.IX Item "\fB\-m\fR \fIdir\fR"
Add dir to the locations of comic modules. Default is /home/benh/share/netcomics.
This option may be repeated to add multiple directories.
.Ip "\fB\-M\fR \fIdir\fR" 4
.IX Item "\fB\-M\fR \fIdir\fR"
Use only the list of locations of comic modules specified on the command line.
.Ip "\fB\-n\fR \fIdays\fR" 4
.IX Item "\fB\-n\fR \fIdays\fR"
Retrieve this number of days of comics, going backwards. Default is 1.
This option may be used in conjunction with \fB\-N\fR.
.Ip "\fB\-N\fR \fIdays\fR" 4
.IX Item "\fB\-N\fR \fIdays\fR"
Start retrieving comics this many days before the currently available date.
If you use \fB\-l\fR to show the comic id's, the 3rd column indicates the number
of days behind a comic is available.  By default, if \fB\-E\fR or \fB\-S\fR are
not specified, then netcomics will retrieve each comic, understanding that
the latest available is that many days ago (according to the number shown
in the 3rd column).  Use this option to push the number of days ago back 
even further.  Default is 0. This option may be used in conjunction with 
\fB\-n\fR, but not with \fB\-S\fR or \fB\-E\fR.
.Ip "\fB\-o\fR" 4
.IX Item "\fB\-o\fR"
Write the webpage on standard out. Cannot specify the number of comics
per page for \fB\-W\fR or \fB\-w\fR if you use this option.
.Ip "\fB\-p\fR \fIurl\fR" 4
.IX Item "\fB\-p\fR \fIurl\fR"
Specify a \s-1URL\s0 to use as a proxy.  Both \s-1HTTP\s0 and \s-1FTP\s0 are supported.
.Ip "\fB\-r\fR \fIrc filename\fR" 4
.IX Item "\fB\-r\fR \fIrc filename\fR"
Specify an alternate name for the user-specific rc file
.Ip "\fB\-s\fR" 4
.IX Item "\fB\-s\fR"
Don't skip bad comics when creating the web page.  This will potentially
cause the web page to be loaded into a browser more slowly, but it will
make it evident exactly which websites don't return proper \s-1HTTP\s0 errors.
.Ip "\fB\-S\fR [\fIdate\fR | \fIdays\fR]" 4
.IX Item "\fB\-S\fR [\fIdate\fR | \fIdays\fR]"
Specify a starting date, or the date that is the specified number of
days prior to today, with which to define the range of days of
comics to retrieve.  May be used in combination with of \fB\-E\fR or
\fB\-n\fR. The date is of the form: M\-D-Y.
.Ip "\fB\-t\fR \fIdir\fR" 4
.IX Item "\fB\-t\fR \fIdir\fR"
Specify the location of html template files. Default is 
/home/benh/share/netcomics/html_tmpl.
.Ip "\fB\-T\fR [\fIdate\fR | \fIdays\fR]" 4
.IX Item "\fB\-T\fR [\fIdate\fR | \fIdays\fR]"
Specify a specific date, or the date that is the specified number of
days prior to today, of comics to retrieve.  This option my be
repeated. The date is of the form: M\-D-Y.
.Ip "\fB\-u\fR" 4
.IX Item "\fB\-u\fR"
Don't download comics, but print URLs on stdout, or if creating a
webpage, make the images be implementing using the URLs.
.Ip "\fB\-v\fR" 4
.IX Item "\fB\-v\fR"
Be a little verbose.
.Ip "\fB\-vv\fR" 4
.IX Item "\fB\-vv\fR"
Be extra verbose
.Ip "\fB\-w\fR[=n]" 4
.IX Item "\fB\-w\fR[=n]"
Create an html file, index.html, for the comics downloaded. Optionally,
n specifies the number of comics to have in each page, where subsequent
html files are named comic#.html.
.Ip "\fB\-nw\fR" 4
.IX Item "\fB\-nw\fR"
Don't create a webpage. Use this option to override the rc file setting
if need be.
.Ip "\fB\-wt\fR \fItitle\fR" 4
.IX Item "\fB\-wt\fR \fItitle\fR"
Specify a title for the web page rather than the default
("Today's Comics From the Web on <\s-1DATE\s0>").  This is useful
for when you download specific comics, and want the title of the
web page to reflect the actual contents.
.Ip "\fB\-W\fR[=n]" 4
.IX Item "\fB\-W\fR[=n]"
Recreate the html file, index.html, from the comics that are in the
directory, as well as any new comics downloaded.  Optionally,
n specifies the number of comics to have in each page, where subsequent
html files are named comic#.html.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Ip "1. Run as a cron job at 7:30am, Monday through Friday, removing the previous day's comics beforehand, and creating a web page. And for Monday, also retrieve Saturday & Sunday's comics." 3
.IX Item "1. Run as a cron job at 7:30am, Monday through Friday, removing the previous day's comics beforehand, and creating a web page. And for Monday, also retrieve Saturday & Sunday's comics."
.Sp
.Vb 2
\&   30 07  *  *  2-5  /home/benh/bin/netcomics -D -w
\&   30 07  *  *  1    /home/benh/bin/netcomics -n 3 -D -w
.Ve
.Ip "2. Same as before, except, for Monday, get Saturday's & Sunday's comics, and for Tuesday, get Monday & Tuesday's.  This is so there isn't such an overload of comics on Monday." 3
.IX Item "2. Same as before, except, for Monday, get Saturday's & Sunday's comics, and for Tuesday, get Monday & Tuesday's.  This is so there isn't such an overload of comics on Monday."
.Sp
.Vb 3
\&   30 07  *  *  1    /home/benh/bin/netcomics -N 1 -n 2 -D -w
\&   30 07  *  *  2    /home/benh/bin/netcomics -n 2 -D -w
\&   30 07  *  *  3-5  /home/benh/bin/netcomics -D -w
.Ve
.Ip "3. Grab Dilbert & Foxtrot comics from the past 30 days, place them in /tmp, and create a web page with a specific title (<\s-1DATE\s0> gets replaced with the name of the month)." 3
.IX Item "3. Grab Dilbert & Foxtrot comics from the past 30 days, place them in /tmp, and create a web page with a specific title (<\s-1DATE\s0> gets replaced with the name of the month)."
.Sp
.Vb 2
\&   netcomics -c "dilbert ft" -n 30 -d /tmp -w -wt 'Dilbert & \e
\&   Foxtrot Comics From the Month of <DATE FORMAT="%b">'
.Ve
.Ip "4. Specify the date range of comics to retrieve to be from Feb 3, 1999 to Feb 6, 1999, and also get comics on March 3, 1998." 3
.IX Item "4. Specify the date range of comics to retrieve to be from Feb 3, 1999 to Feb 6, 1999, and also get comics on March 3, 1998."
.Sp
.Vb 1
\&   netcomics -S 2-3-99 -E 2-6-99 -T 3-3-98
.Ve
.Ip "5. Specify the date range of comics to retrieve to be from Jan 6, 1999 and the 5 days before it.  Get all the comics except Jerkcity and Doodie" 3
.IX Item "5. Specify the date range of comics to retrieve to be from Jan 6, 1999 and the 5 days before it.  Get all the comics except Jerkcity and Doodie"
.Sp
.Vb 1
\&   netcomics -E 1-6-99 -n 6 -C jc -C doodie
.Ve
.Ip "6. Specify the date range of comics to retrieve to be all those that came available three, four, and five days ago." 3
.IX Item "6. Specify the date range of comics to retrieve to be all those that came available three, four, and five days ago."
.Sp
.Vb 1
\&   netcomics -N 3 -n 3
.Ve
.Ip "7. Specify the date range of comics to retrieve to be all those that are dated three, four, and five days ago.  This example is given to show the difference between \-E & \-S  (when given a number) and \-N. All comics downloaded will have the same 3 dates, while in the previous  example, the comics will have varying dates that are determined by the 3rd column in the output of \-l. Note that since many comics aren't available until 2 weeks have past, many of them will not download with this example." 3

NEWS
Changes in release 0.13.1:
* Several bug fixes.
* The option -a ($always_download) when used with -c now only forces
  download of the specified comics.  Failed comics that haven't
  reached the maximum retry limit are also retried.
* New comics in this release:
  bignate     Big Nate
  bigp        The Big Picture
  bz          Bizarro
  bobo        Bobos Progress
  bd          The Boondocks
  bullsnbears Bulls n Bears 
  cwbw        Cartoon Web: BizWit
  cwc         Cartoon Web: Caricatures
  cwet        Cartoon Web: Ecotoon
  kal         Cartoon Web: Kal
  cws         Cartoon Web: Symbolics
  votw        Cartoon Web: Views of the World
  wotw        Cartoon Web: Wit of the World
  cwww        Cartoon Web: WitWomen
  wvotd       Cartoon Web: World View of the day
  cheap       Cheap Thrills Cuisine
  dork        Dork Tower
  exploit     Exploitation Now
  fatjc       Fat Jesus
  fce         Fight Cast Or Evade
  flightdeck  Flight Deck
  grandave    Grand Avenue
  grandclause Grandfather Clause
  jm          James
  lupo        Lupo Alberto
  mrlowe      Mr. Lowe
  pearls      Pearls Before Swine
  pickles     Pickles
  duncan      Raising Duncan
  randolph    Randolph Itch 2:00 AM
  reality     Reality Check
  redrover    Red and Rover
  shirley     Shirley and Son
  soupnutz    Soup to Nutz
  srcwars     Source Wars
* Several comics have been removed either because the artist requested
  it, or the site from which it was being retreaved dropped it.

Major changes in release 0.13:

* Each comics strip's Resource Locator Info (RLI) structure is now
  saved persistently in the spool directory when the comic strip is
  downloaded.  These "rli status files" are reloaded during the next
  invocation of netcomics (unless they're purged) and used to
  determine what comics needs to be retried to be downloaded.  When
  the webpage is recreated, the status files are used to find out
  extra information about the already downloaded comics (author,
  caption, website URLs, etc).  New commandline options: -R -a -q.
* Referers are now handled.  A module may specify its own referer if
  the default (derived from the URL) is not correct.
* The external command specified may now include several macros which
  get replaced.  See the manpage for details.
* New comics in this release: 
  btaf		Bob the Angry Flower
  cotc		Clan of the Cats
  crfh		College Roomies from Hell
  drfun		Dr. Fun
  evjake	Everything Jake
  elf		Elf Life
  gu		Gaming U
  bayside	Life at Bayside
  nukees	Nukees
  sempai	Sempai
  sinf		Sinfest
  wigu		Whenigrowup

Major changes in release 0.12:

* Added the comics: 9 Chickweed Lane, Absurd Notion, Against The
  Grain, Badlands, Bobos Progress, Boonies, Broom-Hilda, Buckets, Cats
  With Hands, Chase Villens, Dilbert Classics, Down To Earth, Dr Katz,
  Eek and Meek, Fair Game, Gibberish, Gil Thorp, Ick, Janes World, The
  Japanese Beetle, K Rat, Lola, Loose Parts, Mary Worth, Meatloaf
  Night, Mickey Mouse, The New Breed, Outtake, PC and Pixel, PS
  Mueller, Penny Arcade, Penny Arcade Bench, Phantom, Player Versus
  Player, Pretzel Logic, Quigmans, Random Shots, Rex Morgan, Ripleys
  Believe It Or Not, Rubes, Small World, Spex and Wally, Staggering
  Heights, Suburban Jungle, Sylvia, Tarzan, Thats Life, Top Of This
  World, Toy Trunk Railroad, WildLife, You Damn Kid, Kudzu, Small
  Grey, and Fat Cats.
* Fixed Y2K bugs.

Major changes in release 0.11:

* Added the comics: After Y2K, Dysentery, Cartoon Web, Bound & Gagged,
  Fred Basset, Herb & Jaamal, Middletons, Pluggers, Motley's Crew,
  Zorro, Liberty Meadows, Teenage Ninja Turtles, Ballard Street, Miss
  Peach, Wee Pals, Raw Material, Agnes, Dr Katz, Outtake, Pcpixel, 9
  Chickweed Lane, Broom-Hilda, Kudzu, Lola, Quigmans, Rubes, Boonies,
  Juilet Jones, Judge Parker, Mark Trail, Henry, Prince Val, New
  Format, Theydoit, Walnut Cove, Tiger, Ripkirby, Hazel, Katzenjammer
  Kids, Mandrake, Pops Place, Popeye, Small Society, Steve Roper,
  Moose Miller, Needhelp, Dick Tracy, Brenda Star, Orphan Annie,
  Gasoline Alley, Soap On A Rope, and World Views
* Added editorial cartoons by: Walt Handelsman,  Bill Day, Chan Lowe,
  Steve Sack, Mike Peters, Dana Summers, Jeff Macnelly, Jack Ohman,
  Don Wright, Dick Locher, Wayne Stayskal
* Changed the default installation location for module files to
  /usr/share/netcomics.
* Added rc file capability.
* Removed Goats for copyright issues.

Major changes in release 0.10:

* Changed from using localtime() throughout the code that downloads
  comics to use gmtime().
* New RLI fields: title, type, author, main, and archives. The field
  called "name" has been depricated. Please read
  Comic_Module-HOWTO.html.
* The created webpages now get extra links included with them: to the
  comic at the website, the comics main page at the website, and the
  website's archives. The body element html template files were
  updated to handle very long headings for each comic.
* The modules have been seperated into their own RPM so that updates
  can be made to them without having to update the the main RPM. The
  name of the RPM is netcomics-data.
* Added the comics: Bliss, The Boondocks, Citizen Dog, The Fusco
  Brothers, Overboard, Tank McNamara, Real Life Adventures, Tradin'
  Paint, Animal Crackers, Bottom Liners, Strange Brew, For Heaven's
  Sake, Heathcliff, & Color Blind.

Major changes in release 0.9:

* The template HTML files now require 8 body_el#.html files where # is
  the number of image files that are handled.  If you have created
  your own HTML template files, be sure to take a look at the ones
  included with this distribution and the document:
  Modify_Webpage_Creation-HOWTO.html
* A new command-line option, -u, has been added.  Using this will
  cause the actual comics to *not* be downloaded.  If you use -W or
  -w, it will make the embedded images link to the comic on its
  website.  If a webpage isn't created, the URLs will be printed on

netcomics.pl
#!/usr/bin/perl -w
#-*- tab-width: 4 -*-
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#This script requires the following CPAN packages:
#libwww-perl, HTML-Parser, Image-Size
#
#To add your own comic lib modules, first visit the comic strip resource at:
#http://comics.kurle.com/resource/
#to find the comic you want.

$0 =~ s,.*/,,;  # use basename only

use POSIX;
use strict;
use Carp;
use Netcomics::MyResponse;
use Netcomics::MyRequest;
use Netcomics::ExternalUserAgent;
use Netcomics::RLI;
use Netcomics::HTML;
use Netcomics::Factory;
use Netcomics::Config;
use Netcomics::Util;

# Global Variables
my $script_name = "netcomics";

my $conf = Netcomics::Config->new($script_name);
$conf->processARGV(@ARGV);

$| = 1; #autoflush on STDERR & STDOUT

#Persistent storage of rli info depends on Data::Dumper
my $data_dumper_installed = requireDataDumper();

my $factory = Netcomics::Factory->new($conf);
$factory->setup;

$factory->list_comics(), exit(0) if ($do_list_comics);

#get_comics returns a list of comics which is used to help determine
#if a comic in the directory was just downloaded or not.
#dump_rli(\@rli) if $single_rli_file;
my @rli = $factory->get_comics();
my @comics = $factory->files_retrieved();
my @existing_rli_files = $factory->existing_rli_files();
my $get_current = $factory->get_current();
if ($remake_webpage) {
	print "Scanning $comics_dir for comics with no associated status file.\n"
		if $extra_verbose;
	opendir(DIR,$comics_dir) || die "could not open $comics_dir: $!";
	my @files = readdir(DIR);
	closedir(DIR);
	#sort so that if there's a multifile comic and regular comic associated
	#with the same comic-date, that the multi-file comics can be processed
	#first.
	@files = sort(grep(/(xpm|gif|jpe?g|tiff?|png)$/,@files));

  RMW:
	while (@files) {
		my $file = shift(@files);
		#although we could check to see if the file is in @comics or
		#@existing_rli_files right away, we have to first check through
		#all of the files associated with this one if it is a multi-file comic
		my ($title,$date,$type,$num) = parse_name($file);
		my $name = "$title-$date";
		my $rli = {'title' => $title,
				   'name' => $name,
				   'time' => date_from_filename($file),
				   'status' => 1,
				   'type' => $type,
				   'file' => [$file]
				   };
		if (defined($num)) {
			#this is part of a multi-file comic
			#recreate the file array because the one we're looking
			#at may not be image #1.
			$rli->{'file'}->[$num - 1] = $file;
			my @multifiles = grep(/^$name(-\d+)?\.\w+$/,@files);
			@files = grep(!/^$name(-\d+)?\.\w+$/,@files);
			for (@multifiles) {
				my $file = $_;
				#make sure that none of the files in the multi-file comic
				#have an rli status file
				if (grep(/^$file$/,@comics) ||
					grep(/^$file$/,@existing_rli_files)) {
					print STDERR "Warning: $name has some stale file(s).\n"
						if $verbose;
					next RMW;
				}
				#add this file to the $rli's file list
				my ($title,$date,$type,$num) = parse_name($file);
				$rli->{'file'}->[$num - 1] = $file;
			}
		}
		next if grep(/^$file$/,@comics) ||
			grep(/^$file$/,@existing_rli_files); #skip this file 
		#no associated rli status file, generate our own rli hash for the file.
		print STDERR "Warning: $name has no status file; some " .
			"info about it may be lost.\n" if $verbose;
		push(@rli,$rli);
	}
}

#check to see if any comics were downloaded or are downloadable
#simply to give a message to the user if verbose.
if ($verbose) {
	my $good = 0;
	foreach (@rli) {
		$good = 1, last if $_->{'status'};
	}
	unless ($good) {
		my $m = $dont_download ? "are downloadable" : "were downloaded";
		print "\nNo comics $m.\n";
	}
}

if ($make_webpage) {
	#Create the HTML object.
	my $HTMLpage = Netcomics::HTML->new("Webpage");

	# Create the webpage.
	unless ($separate_comics) {
		$HTMLpage->create_webpage(@rli);
	} else {
		$HTMLpage->create_webpage_set(\@rli, \@selected_comics);
	}
} else {
    my $user_specified_no_comics = 
	($user_specified_comics && ! @selected_comics) ? 1 : 0;
    #print filenames or urls.
	foreach (@rli) {
		my $rli = $_;
		if ($rli->{'status'} =~ /[12]/) {
			my $time = $rli->{'time'};
			if ($get_current) {
				if (defined($rli->{'behind'})) {
					#get what the current specified date was from days behind
					$time +=  $rli->{'behind'}*3600*24;
				} elsif (defined($factory->days_behind($rli->{'proc'}))) {
					#backwards compatibility for rlis created with code that
					#didn't save the days behind in the rli.
					$time += $factory->days_behind($rli->{'proc'})*3600*24;
				} else {
					#this must be an old rli file not created with code
					#that saved the days behind info in the rli.
					$time = "";
				}
			}
			if (!$user_specified_comics || $user_specified_no_comics ||
				(grep(/^$rli->{'proc'}$/, @selected_comics) &&
				 grep(/$time/, @dates))) {
				print "$rli->{'title'} (" . 
					strftime("%a, %x",gmtime($rli->{'time'})) . "):\n" 
						if $verbose;
				#dont_download is set when user requests urls only
				if ($dont_download) {
					print join("\n",@{$rli->{'url'}})
						if defined $rli->{'url'};
				} else {
					print join("\n",map {"$comics_dir/$_"} @{$rli->{'file'}})
						if defined $rli->{'file'};
				}
				print "\n";
			}
		}
	}
}
#END of main

